examples/auction/src/lib.rs

```diff
@@ -1,208 +1,201 @@
-) -> InitResult<State<S>> {
+    _state_builder: &mut StateBuilder<S>,
+) -> InitResult<State> {
+    // Getting input parameters
     let parameter: InitParameter = ctx.parameter_cursor().get()?;
+    // Creating `ViewableState`
     let viewable_state = ViewableState {
-        auction_state: AuctionState::NotSoldYet,
-        highest_bid:   Amount::zero(),
-        item:          parameter.item,
-        expiry:        parameter.expiry,
+        auction_state:  AuctionState::NotSoldYet,
+        highest_bidder: None,
+        item:           parameter.item,
+        end:            parameter.end,
     };
+    // Creating `State`
     let state = State {
         viewable_state,
-        bids: state_builder.new_map(),
     };
     Ok(state)
 }
 
-/// Receive function in which accounts can bid before the auction end time
+/// Receive function for account addresses to place a bid in the auction
 #[receive(contract = "auction", name = "bid", payable, mutable)]
 fn auction_bid<S: HasStateApi>(
     ctx: &impl HasReceiveContext,
-    host: &mut impl HasHost<State<S>, StateApiType = S>,
+    host: &mut impl HasHost<State, StateApiType = S>,
     amount: Amount,
 ) -> Result<(), BidError> {
+    let balance = host.self_balance();
     let state = host.state_mut();
+    // Ensure the auction has not been finalized yet
     ensure!(
         state.viewable_state.auction_state == AuctionState::NotSoldYet,
         BidError::AuctionFinalized
     );
 
     let slot_time = ctx.metadata().slot_time();
-    ensure!(
-        slot_time <= state.viewable_state.expiry,
-        BidError::BidsOverWaitingForAuctionFinalization
-    );
+    // Ensure the auction has not ended yet
+    ensure!(slot_time <= state.viewable_state.end, BidError::BidsOverWaitingForAuctionFinalization);
 
+    // Ensure that only account addresses can place a bid
     let sender_address = match ctx.sender() {
         Address::Contract(_) => bail!(BidError::ContractSender),
         Address::Account(account_address) => account_address,
     };
-    let mut bid_to_update = state.bids.entry(sender_address).or_insert(Amount::zero());
-    *bid_to_update += amount;
-    // Ensure that the new bid exceeds the highest bid so far
-    ensure!(*bid_to_update > state.viewable_state.highest_bid, BidError::BidTooLow);
-
-    state.viewable_state.highest_bid = *bid_to_update;
 
+    // Ensure that the new bid exceeds the highest bid so far
+    ensure!(amount > balance, BidError::BidTooLow);
+
+    if let Some(account_address) = state.viewable_state.highest_bidder.replace(sender_address) {
+        // Refunding old highest bidder;
+        // This transfer (given enough NRG of course) always succeeds because the
+        // `account_address` exists since it was recorded when it placed a bid.
+        // If an `account_address` exists, and the contract has the funds then the
+        // transfer will always succeed.
+        if host.invoke_transfer(&account_address, balance).is_err() {}
```

Comment:
This is a strange way of doing it. if you want to ignore the error then the more canonical way would be
```rust
let _ = host.invoke_transfer(...)
```
I don't think I've ever seen a dummy if statement like you have it.

However I don't like silently ignoring errors (even if they should not happen). It is better to do
```
host.invoke_transfer(...).unwrap_abort()
```
which will terminate the contract in case the impossible happens.

Reply:
If I understand correctly “host.invoke_transfer(...).unwrap_abort()” is equivalent to “bail!(...)”. Meaning it will revert the tx in case the CCD transfer fails, isn’t it?

I am worried about [this scenario](https://consensys.github.io/smart-contract-best-practices/attacks/denial-of-service/). I suspect that people will use our auction example one day and change it to allow smart contracts to take part in the auction (in that case it would be a problem if we revert the “host.invoke_transfer(...)” on an error). The ideal solution would be a pull-over-push pattern but I don’t want to complicate the auction example even further, therefore I changed it to this option now. Please let me know if that is ok, or if you want to proceed differently.

let _ = host.invoke_transfer(...)